### ISSUE TITLE
Tripal 3 Chado preparation issue when reloading a Chado dump

### ISSUE THREAD

**Issue by guignonv** [tripal/tripal#322]
* Tripal Version: 7.x-3.x
* Drupal Version: not related
* PostgreSQL Version: not related
* PHP Version: not related

Note: this is a minor issue as it should only happen to advanced users manipulating the database directly.

When you install and initialize Chado with Tripal (job) using "Prepare Chado" interface, materialized views are created correctly and everything is fine. However, if you clear the Chado schema in order to load a new one and want to re-"Prepare Chado" again, the materialized views may not be present in the dump while they are recorded as existing on the Drupal (public schema) side. Then, many errors will appear everywhere and it is not possible to "Prepare Chado" again.
Typical error messages: 'Please update the database using "drush updatedb" before continuing' and 'Materialized view, db2cv_mview, already exists. Skipping creation.'.

The workaround I used to solve that issue was to clear "tripal_mviews" table: "DELETE FROM tripal_mviews;" and re-run the "Prepare Chado" process.

Proposition: in tripal/tripal_chado/api/tripal_chado.mviews.api.inc around line 55, the presence/absence of a materialized view table is checked using `'SELECT mview_id FROM {tripal_mviews} WHERE name = :name'` to see if it is recorded in the tripal_mviews table. However, the table may not exist for real. A different approach should be used. What about:
```
SELECT EXISTS (
   SELECT 1
   FROM   information_schema.tables 
   WHERE  table_schema = :chado_schema
   AND    table_name = :name
   );
```
But the synchronization between existing materialized views and whats is recorded in the tripal_mviews table should be also checked in a way or another... Maybe a kind of "repair" function for materialized views should be added and interfaced?

**Comment by laceysanderson** [comment]
I agree that a different means of checking this is a good idea! Perhaps using [chado_table_exists](http://api.tripal.info/api/tripal/tripal_chado%21api%21tripal_chado.schema.api.inc/function/chado_table_exists/3.x)?

We could alleviate the need for an interface if we added the following at line 55:
```php
// Check that the materialized view actually exists and if not, 
// remove the entry from tripal_mviews.
if ($mview_id AND !chado_table_exists($name)) {
   db_query(
    'DELETE FROM {tripal_mviews} WHERE mview_id = :id',
    array(':id' => $mview_id))->execute();
  $mview_id = FALSE;
}
```
This way, the mview table is cleaned up when it's checked for a view that has gone missing. No need for an admin to trigger a repair and the mview can be added in the same call.

Thoughts?

**Comment by spficklin** [comment]
@laceysanderson I like your fix.  I'll take care of the pull request on this.

**Comment by laceysanderson** [comment]
This should be fixed by PR #404 as confirmed by two reviewers. If this doesn't fix the issue for you @guignonv please re-open the issue!
